Reworked Abins DMOL3 parser

ajjackson opened the pull request:
**Description of work.**

Addresses #34341 : the DMOL3 parser currently scrambles the eigenvector matrix when the data takes more than one block of 9 columns to present. (i.e. when more than 9 eigenvectors are printed). This data is used to determine intensities in Abins. In some cases the resulting error may not be obvious, as frequencies will still be correct but peak intensities are not.

The problem was detected when a user file _happened_ to yield a large block of zeros when read incorrectly, causing a divide-by-zero error in a normalisation step that should never see a total of zero.

There are a couple of other odd features of the parser that can be addressed at the same time:
- a normalisation constant that is stored as state on the parser and will keep accumulating if the file is read repeatedly. It is unclear if this does any long-term harm, as another normalisation step follows. But there should not be state changes resulting from repeated reader calls from the same file: that is surprising behaviour.
- the second normalisation uses the default Mantid masses, but these may not be consistent with those used in the eigenvector calculation? It would make more sense to normalise by the reported masses, which reflect any isotopic substitutions. _(Update: indeed, it seems that good results come when the mass weighting is dropped entirely. Presumably this was already baked into the DMOL output.)_

The root cause of the eigenvector problem is that it was difficult to inspect the output eigenvectors and compare them with the input file, because they were re-ordered and normalised at read-time. The new implementation should separate the reading and post-processing so that it is easier to sanity-check the intermediate results. By making the testing more atomic we need rely less on file-based tests to cover all the edge cases.

The new unit tests include some literal text blocks. This really helps with understanding that the test is doing the right thing, but if preferred that data could be pushed into a JSON file or something for a more compact (but less standalone) unit test file. I don't think it's obnoxiously large at this point.

I'm happy to report that the actual parser hasn't got any bigger. The net lines added by this PR are mostly testing-related!

**Report to:** Stewart Parker

**To test:**
The reworked parser should have better unit tests than the original version.

The existing numerical unit tests (on the whole loader, against realistic files) have been updated. A problem remains that these are not well validated against reference data.

A validation check against CRYSTAL is shown below (but this is for quite a simple system and could miss some nuance)

Fixes #34341 

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

ajjackson committed:
Modernise Abins TextParser base class

Significant work is needed on the DMOL3 parser; as it uses this class,
take the opportunity to clean up a bit.

- Remove redundant (default) base class and __init__
- Use type-hinting to make API easier to read
- Allow positional arguments where there would be little ambiguity
- Mark staticmethods and classmethods; note that in fact this is all
  the methods, so this class is really just a namespace. We could
  consider migrating them to module-level functions and removing an
  unnecessary layer of state.

ajjackson committed:
Abins AbInitioLoader typing cleanup

Clean up type hints and docs to make this class a bit easier to understand

ajjackson committed:
Abins DMOL3 parser: Modernise file and test lattice vectors

- No need to instantiate TextParser now, it's only classmethods and
  staticmethods

- _read_lattice_vectors can be a classmethod too

- This allows direct testing of the method with a sample data
  block. With this approach we can have more atomic unit tests that
  don't rely on reading a whole file to cover every corner of the
  behaviour.

- (Actually we are not too worried about the _read_lattice_vectors
   part of the code, but this will be a good model for properly
   developing and testing other parts of the parser.)

ajjackson committed:
Abins DMOL3 parser: simplify lattice vector read

- We don't need a private function to do multiplication in one place.

- We can also let numpy take care of string conversion and vectorise
  the unit conversion

ajjackson committed:
Abins dmol3 parser - more functional approach to lattice vectors

Pure functions are easier to test and understand than ones that work
through side-effects. Return lattice vectors directly rather than take
a dict to mutate; have the calling function manage dictionary assignment.

ajjackson committed:
Abins parser cleanup: refactor save-excursion to context manager

This is a repeating pattern that appears in several Abins parsers, and
twice in the DMOL3 parser: save the current file position to a
variable, and after various readline() and seek() operations restore
the position so the next parsing stage has a clean/predictable
starting point.

This is a bit vulnerable to variable-name clashes or editing as the
idiom requires separate statements at each end of a function and an
otherwise-unused variable to sit in the function namespace.

Here we move to a context manager, so the scope is clearly indicated
by indentation and we can rely on a safe exit if the block terminates
unexpectedly. This is located on TextParser so it can be used by other
Abins parsers (but that can be left to a later PR if necessary while
the urgent DMOL3 changes are prioritised.)

ajjackson committed:
Abins DMOL3 parser: further refactor and tests of positions reading

Separate the isotope management and dict mutation from the actual
parsing method. This allows simpler testing that output looks as
expected, and reduces both function arguments and side-effects.

Other code is made less verbose by use of f-strings, numpy features
and multiple-assignment.

I'm not sure why the long line of ------- may be needed as an
alternate end token, but a shorter line should also match and keeps
the source file tidy.

ajjackson committed:
Abins DMOL3 parser: test mass reader

The masses reader can also be a clean staticmethod and tested
independently. This lets us craft fake data to check the alternate
termination condition (i.e. end of file.)

ajjackson committed:
Abins DMOL3 parsing: Rewrite import of phonon mode data

- The transformation from DMOL3 data tables is now done in several
  steps. This makes it easier to verify that each transformation is
  handled as expected:

  - reading data rows in blocks (where the final block has a variable
    number of columns)
  - reshaping the data array from 2-D with (x,y,z) interleaved to 3-D
  - mass-weighting
  - normalisation

- Classmethods tests are added to sanity-check the block reading, but
  not all cases are covered yet. The test is moved to a new file and
  actually runs now.

- The output of normalisation is checked for all parsers by adding a
  method to KpointsData and calling this check in tests. It seems to
  be consistent with existing results for other parsers.

- However it is difficult to be 100% confident in the normalisation
  _process_ without deeper comparison with existing methods and/or other
  codes. It is, at least, much simpler than before.

As expected, the phonon modes have changed for the large crystal test
case. It is more surprising that they have changed for the small
molecule test case. The results should be checked carefully against
another code.

ajjackson commented:
Update: I think the new parser is a lot more testable and understandable. 
It returns eigenvectors that are different to the existing ones, as expected. I would like to take a bit more time to really sanity-check these results against the other data formats. Unfortunately I will be on annual leave for the next week; I appreciate this reduces the likelihood of getting the fix into release-next. We shall see what happens!

ajjackson committed:
Abins DMOL3 parser: more unit testing of eigenvalue reader

Test the eigenvalue block reader
- Use inline data instead of external files so we can easily
  sanity-check the expected values for each case. (If desired, the
  data could all go into a JSON file or something...)
- Check effect of different line lengths so we are confident that
  parsing and concatenation work together correctly

martyngigg commented:
Hi @ajjackson,

Do we think we might be able to get this fix in? If it helps at all I've just removed the need to support RHEL7 (python 3.6)  so if we rebase this on `release-next` that will be one less thing have to worry about.

ajjackson commented:
That's great news! I'll run the rebase in a minute then.

**Executive summary: code and unit tests already seem good to go, but validation is lacking**

I am currently rather conflicted on this fix: on the one hand, the new parser is clearly better-written, better-tested and removes a demonstrable problem.

But empirically the improvement is less obvious. For the small molecule (methane) case which I had previously believed to be unaffected by the scrambling, the DMOL3 results seem to get _worse_ compared to a CRYSTAL17 calculation of the same system. After a day or two digging through the numbers I am fairly convinced that the small-molecule case was also wrong and should also be fixed by the new parser; yet empirically it appears that somehow it has got worse.

This may be a coincidence related to other differences in the results between calculators; "fortuitous cancellation of errors" being painfully undone. From the end of November I will begin host a graduate project that aims to generate a larger test set so that this kind of thing can be investigated much more systematically.

Or perhaps there is a mistake in the new code.

I want to avoid the situation that we announce a "fix" in release that immediately needs revising in nightly or the next point release. The resulting confusion could be just as -or more- harmful than the bug, as it might encourage more people to use a broken feature...

I will try to think of another validation test that can _quickly_ be used to clarify the situation.

ajjackson committed:
Abins DMOL3 import: Remove mass weighting

Mass weighting was originally included in this update to the parser as
it was already there. However, it's not entirely clear that is should
be.

Removing the mass weighting seems to give much better agreement with
b_traces from CRYSTAL and CASTEP calculations, so we shall assume it
is the correct behaviour.

ajjackson commented:
I think I've got to the bottom of it. The sqrt(mass) factor applied to the DMOL3 eigenvectors, which I imitated from the existing parser, should not be there. This did have some kind of fortuitous cancellation of errors with the eigenvector scrambling, but now that _both_ things are fixed we get convincing agreement with a CRYSTAL17 calculation for the same system.

Of course the frequency difference remains, but the intensities now match the CRYSTAL calculation quite closely. By visual inspection the traces of the displacement tensors (i.e. `b_traces` in Abins) are also similar to the results from CASTEP. (There's no point in including the CASTEP results on this plot though as that calculation is dominated by spurious rotational modes which were removed in the open-boundary codes.)

![dmol-vs-crystal](https://user-images.githubusercontent.com/2511383/194092876-debabbc4-ea06-4bd5-8418-29ec4f7080f1.png)

ajjackson committed:
Abins DMOL3: new unit test data

Reference files for DMOL3 importer. The loader has changed quite a
lot, with the following expected to change outputs:
- ordering of displacements corrected
- mass weighting of displacements removed

ajjackson commented:
**Executive summary: I'm happy with the maths for now**

I have run another test calculation and inspected the data and am now fairly confident that
- the raw data from DMOL3 are normalised eigenvectors (also known as "polarisation vectors")
- it is correct to convert these to atomic displacements by dividing these through by sqrt(mass) as Abins does (indirectly) in powdercalculator.py

This puts them on a similar footing to the data from CASTEP, which also does not apply mass weighting in the parser.

CRYSTAL is a bit different; the raw data are atomic displacements with the amplitude of the zero-point vibrational mode. The parser converts these to the expected normalised polarisation vectors. (The exact means by which it does this remains mysterious to me, but is no longer today's problem.)

ajjackson committed:
Modernise abins.input.tester

- Clean up commented line
- assertEqual(True, ...) should be assertTrue
- numpy.testing.assert_allclose gives much more useful error messages
  than assertTrue(np.allclose(...)) (shows data and gives stats)

AnthonyLim23 commented on scripts/abins/input/abinitioloader.py:
```diff
@@ -138,14 +142,14 @@ def load_formatted_data(self):
     # Internal method for use by child classes which read ab initio phonon data
-    def _rearrange_data(self, data=None):
+    @staticmethod
+    def _rearrange_data(data: dict) -> abins.AbinsData:
```
is it ok that we have removed the default value for data?

ajjackson replied:
I think so; it's useless to call this function with `data=None` and will immediately lead to a KeyError.

A lot of functions in Abins were written with non-functioning `None` default parameters; I think the intent behind this was to implement mandatory keyword arguments. Python has a proper syntax for that since Python 3 (https://peps.python.org/pep-3102/) but at the time Abins still supported Python 2.7.

I think the intent behind using a pseudo-required-kwarg was to improve readability by using explicitly named arguments in function calls. But I don't agree that adding the word "data" is beneficial for a function like this one that only takes one argument, so while adding the type hints it seems a good chance to simplify things as well.

The positional argument will still be backward-compatible with any code still using `data=...`; it is not a [positional-only parameter](https://peps.python.org/pep-0570/).

AnthonyLim23 commented on scripts/abins/input/abinitioloader.py:
```diff
@@ -159,7 +163,7 @@ def _rearrange_data(self, data=None):
         return abins.AbinsData(k_points_data=k_points, atoms_data=atoms)
 
-    def save_ab_initio_data(self, data=None):
+    def save_ab_initio_data(self, data: dict) -> None:
```
same as above

ajjackson replied:
As above: actually running the function with `None` will raise an error either way. This should make it clearer that you aren't supposed to do that!

AnthonyLim23 commented on scripts/abins/input/abinitioloader.py:
```diff
@@ -192,10 +196,20 @@ def get_formatted_data(self):
     @staticmethod
-    def check_isotopes_substitution(atoms=None, masses=None, approximate=False):
+    def check_isotopes_substitution(atoms: dict,
+                                    masses: Sequence[float],
```
atoms and masses have lost their defaults. Is that ok?

AnthonyLim23 commented on scripts/abins/input/dmol3loader.py:
```diff
@@ -41,245 +55,233 @@ def read_vibrational_or_phonon_data(self):
+            else:
+                vectors = [file_obj.readline().split() for _ in range(3)]
+                unit_cell = np.asarray(vectors, dtype=FLOAT_TYPE)
+                unit_cell *= ATOMIC_LENGTH_2_ANGSTROM
```
why 2 angstrums?

ajjackson replied:
I think 2 is a 1337 abbreviation for "TWO" in this case. Agree that it's confusing, could change it.

Technically that would be an API break if anyone is importing that from the Abins modules

AnthonyLim23 replied:
you could for this file do

`` 
import ATOMIC_LENGTH_2_ANGSTROM as BOHR_TO_ANGSTROM
```

or similar. Then it is not an API break, but in this code is more readable.

ajjackson replied:
Right, I much prefer referring to Bohr explicitly anyway. It might as well happen as an alias in the constants file, and there can be a note of the intended deprecation somewhere for when the version number is next bumped.

AnthonyLim23 commented on scripts/abins/input/dmol3loader.py:
```diff
@@ -41,245 +55,233 @@ def read_vibrational_or_phonon_data(self):
+        for line in file_obj:
+            if line.decode('utf8').strip():
+                file_obj.seek(last_line)
+                break
```
could this just be a return instead of a break?

ajjackson replied:
It could, if you think that would be clearer? I'm a bit wary of `return` statements in the middle of functions but this seems like a good place to use one.

AnthonyLim23 replied:
I am always wary of breaks. I find it takes more thought to work out where the code goes to next.

AnthonyLim23 commented on scripts/abins/input/dmol3loader.py:
```diff
@@ -41,245 +55,233 @@ def read_vibrational_or_phonon_data(self):
-        :param file_obj: file object from which we read
-        :param data: Python dictionary to which k-point data should be added
+        frequencies, displacements = cls._read_mode_block(file_obj)
+        frequency_blocks = [frequencies]
```
could we not initialise the frequency_blocks and displacement_blocks here and then do all of the assignments in the while loop? Then if anything changes in the future its one edit and not 2.

ajjackson replied:
Good catch. Can only assume it made more sense in an previous iteration 😅

AnthonyLim23 commented on scripts/abins/input/dmol3loader.py:
```diff
@@ -41,245 +55,233 @@ def read_vibrational_or_phonon_data(self):
+            # if so number of modes was not multiple of 9 and extra columns
+            # should be removed before assembling blocks to full array
+            if np.all(np.isnan(displacement_blocks[-1][:, column])):
+                displacement_blocks[-1] = displacement_blocks[-1][:, :column]
```
do we not need to change the frequencies as well?

ajjackson replied:
The frequencies are read using readline.split() so always come in at the right length.

The displacements are read using a regular expression that matches 9 groups, 8 of them optional. The unused optional matches (i.e. empty columns) turn up as `None` in `match.groups()`, get converted to `NaN` by numpy.asarray() and then get cleaned up here.

AnthonyLim23 replied:
could we have a comment explaining that?

ajjackson replied:
Added one, does it look ok?

AnthonyLim23 commented on scripts/abins/input/dmol3loader.py:
```diff
@@ -41,245 +55,233 @@ def read_vibrational_or_phonon_data(self):
+
+        Displacements are also normalised here. Note that owing to a different
+        scaling/reporting convention a sqrt(mass) factor is _not_ applied as
+        with e.g. CRYSTAL17 output.
```
```suggestion
        with other output (e.g. CRYSTAL17).
```

ajjackson replied:
Gotcha. Will word in slightly more detail than this.

AnthonyLim23 replied:
I'm just being fussy with English.

ajjackson committed:
Abins DMOL3 parser: code review suggestions

Thank you @AnthonyLim23

- alias ATOMIC_LENGTH_2_ANGSTROM to something less misleading
- "return" is clearer than "break" in this particular function
- all blocks can be read inside loop
- write docstring in proper English

AnthonyLim23 commented on scripts/abins/input/textparser.py:
```diff
@@ -87,11 +104,13 @@ def file_end(self, file_obj=None):
+
         :param file_obj: file object from which we read
-        :param msg: list with messages which end kpoint block.
+        :param msg: list of messages which can end kpoint block.
```
```suggestion
        :param msg: list of messages that can end kpoint block.
```

AnthonyLim23 commented on scripts/abins/kpointsdata.py:
```diff
@@ -129,6 +130,17 @@ def get_gamma_point_data(self):
+            if not np.allclose(np.ones(displacements.shape[1]),
+                               norm(norm(displacements, axis=0), axis=1)):
+                return False
+        else:
```
this else does not match up with anything

ajjackson replied:
It's valid Python, for-else is used for a statement that should only run if the loop is not otherwise broken. https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

It should really be used with `break` rather then `return`, though: in thise case it can be removed.

AnthonyLim23 commented on scripts/test/Abins/AbinsLoadDMOL3InternalsTest.py:
```diff
@@ -0,0 +1,257 @@
+                      [2.42243481736875, 8.68442407988897, 6.67097246055480],
+                      [-2.34542583215868, 10.72867081788549, 2.94943237944894],
+                      [0.04185270935330, 2.59517848080392, -1.80983568538814]]
+                     ) * 0.52917721067  # Convert Bohr to Angstrom
```
we should import the conversion factor (forces the same precision and is cleaner). Otherwise we should just report the values in Angstroms.

ajjackson replied:
I worry that this will miss the case that something goes wrong with `abins.constants`, but I guess that will still come up in the numeric tests!

AnthonyLim23 replied:
either that or do a test to check that the constant has a specific value

ajjackson replied:
Gone with import: meaningful names > comments

ajjackson committed:
Clarify docstring in scripts/abins/input/textparser.py

Co-authored-by: Anthony <anthony.lim@stfc.ac.uk>

ajjackson committed:
DMOL3 parser: more review comments

Thanks @AnthonyLim23

- Comment more on need to clean up eigenvectors but not
  frequencies.

- Remove unnecessary "else" clause

- Import unit conversion factor rather than hard-code

AnthonyLim23 commented:
I am happy to approve this once the tests pass

ajjackson commented:
There is actually systemtest data for the same system calculated with CASTEP. Making a three-way comparison... we can _maybe_ argue that the new parser intensities agrees better in the high frequencies, but it's a weak claim. The difference in calculated frequencies between CASTEP and DMOL (not the fault of the parsers...) is bigger than the difference between DMOL parsers in the low-mid frequency range.

![dmol-castep-systemtest](https://user-images.githubusercontent.com/2511383/195405001-5fb7cc51-dae3-4cc9-b209-7e7a75a29ab1.png)

I hope to make some more rigorous cross-code comparisons next year, working with an SCD graduate. Until then this is the best we have.

ajjackson committed:
Update DMOL3 systemtest data

As discussed in https://github.com/mantidproject/mantid/pull/34418

- frequencies agree with old data, as expected
- Updated intensities appear to agree better with CASTEP data, but
  these are different calculations so comparison should be taken with
  pinch of salt

ajjackson commented:
Aha, the intensity improvement is more convincing when we simplify the spectrum by only plotting the order-1 Fluorine atom contributions

![dmol-castep-o1-systemtest](https://user-images.githubusercontent.com/2511383/195406634-181005fa-663c-4fe4-8a09-fb0c1700063f.png)

ajjackson committed:
Abins DMOL3 parser fix: changelog update

ajjackson commented:
Release note added, are there any issues making sure it gets into the formatted docs at this stage?

sf1919 commented:
The release note will be added to the appropriate page manually as part of final cleanup by the release editors once the PR is merged in. I'm not imaging there will be any issues but we'll be in touch if anything crops up.

AnthonyLim23 reviewed (approved):
Looks good to me and the code no longer crashes when reading some input files
